ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/InferredTypeHintStage.fs

```diff
@@ -0,0 +1,175 @@
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Util
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Util
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Highlightings
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl
+open JetBrains.ReSharper.Psi
+open JetBrains.ReSharper.Psi.Naming
+open JetBrains.ReSharper.Psi.Naming.Impl
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.ReSharper.Psi.Util
+open JetBrains.ReSharper.Feature.Services.Daemon
+open JetBrains.ReSharper.Plugins.FSharp.Daemon.Cs.Stages
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open FSharp.Compiler.SourceCodeServices
+
+type LocalReferencePatternVisitor(fsFile: IFSharpFile, highlightingContext: TypeNameHintHighlightingContext, namingPolicyProvider, nameParser) =
+    inherit TreeNodeVisitor<IHighlightingConsumer>()
+
+    let isTypeEvidentFromVariableNamePrefix (typ: IType) (variableNameParts: string[]) =
+        if not (typ.IsBool()) then false else
+
+        if variableNameParts.Length > 0 then
+            let prefix = variableNameParts.[0].ToLowerInvariant()
+            prefix = "has" || prefix = "is"
+        else
+            false
+
+    let isEvidentFromVariableName (fsType: FSharpType) variableName =
+        if not highlightingContext.HideTypeNameHintsWhenTypeNameIsEvidentFromVariableName then false else
+
+        let typ = fsType.MapType(Array.empty, fsFile.GetPsiModule())
+        if not (typ.IsValid()) then false else
+
+        let variableNameParts = NamesHelper.GetParts(nameParser, namingPolicyProvider, variableName)
+        if isTypeEvidentFromVariableNamePrefix typ variableNameParts then true else
+
+        match typ.GetTypeElement() with
+        | null -> false
+        | typeElement ->
+
+        let typeName = typeElement.ShortName
+        String.Equals(typeName, variableName, StringComparison.OrdinalIgnoreCase) ||
+        not (String.IsNullOrEmpty(typeName)) &&
+        not (String.IsNullOrEmpty(variableName)) &&
+        NamesHelper.IsLike(variableNameParts, NamesHelper.GetParts(nameParser, namingPolicyProvider, typeName))
+
+    let isTypeOfPatternEvident (pattern: ISynPat) =
+
+        // v-- not evident
+        // x::y::z
+        //    ^--^-- evident
+        let tuplePat = TuplePatNavigator.GetByPattern pattern
+        if isNotNull tuplePat then
+            let consPat = ConsPatNavigator.GetByPattern1 tuplePat
+            if isNull consPat then false else
+
+            // Are we nested inside another ConsPat?
+            let parentTuplePat = TuplePatNavigator.GetByPattern consPat
+            if isNull parentTuplePat then tuplePat.Patterns.IndexOf pattern <> 0 else
+            isNotNull (ConsPatNavigator.GetByPattern1 parentTuplePat)
+
+        else
+
+        //  v-- not evident
+        // [x; y; z]
+        //     ^--^-- evident
+        let listOrListPat = ArrayOrListPatNavigator.GetByPattern pattern
+        if isNotNull listOrListPat then
+            listOrListPat.Patterns.IndexOf pattern <> 0
```

Comment:
It would be a bit cheaper to get `PatternsEnumerable` instead of getting a collection of all patterns and to do a reference equality check with the first one.